test.py

```diff
@@ -0,0 +1,595 @@
+            framework = fwinfo.get_framework(framework_name)
+            if framework.popen:
+                framework.popen.terminate() # does nothing if already completed
+    return all_ok
+
+def run_tests(run_attrs, repo_root):
+    logger.info("cluster_teardown policy: %s", run_attrs.cluster_teardown)
+    try: # all clusters are set up inside this try
+        all_passed = False
+        if run_attrs.parallel:
+            logger.debug("Running m ulticluster test run")
+            all_passed = _multicluster_linear_per_cluster(run_attrs, repo_root)
+        else:
+            all_passed = _one_cluster_linear_tests(run_attrs, repo_root)
+    finally:
+        if run_attrs.cluster_teardown == "always":
+            teardown_clusters()
+        elif run_attrs.cluster_teardown == "success-only" and all_passed:
+            teardown_clusters()
+
+        for cluster in clustinfo._clusters:
+            logger.debug("Cluster still running: url=%s id=%s auth_token=%s",
+                         cluster.url, cluster.cluster_id, cluster.auth_token)
+
+
+def _setup_strict(framework, cluster, repo_root):
+    "Do cruft required to create the roles for the framework"
+    security = os.environ.get('SECURITY', '')
+    logger.info("SECURITY set to: '%s'", security)
+    if security == "strict":
+        logger.info("running %s role strict setup script(s)", framework.name)
+        perm_setup_script = os.path.join(repo_root, 'tools', 'setup_permissions.sh')
+
+        custom_env = os.environ.copy()
+        custom_env['CLUSTER_URL'] = cluster.url
+        custom_env['CLUSTER_AUTH_TOKEN'] = cluster.auth_token
+
+        for script_num in (1, 2):
+            role_arg = '%s%s-role' % (framework.name, script_num)
+            cmd_args = [perm_setup_script, 'root', role_arg]
+
+            completed_cmd = subprocess.run(cmd_args, env=custom_env)
+            if completed_cmd.returncode != 0:
+                msg = "%s invocation returned failure.  FAIL"
+                logger.info(msg, " ".join(cmd_args))
+                raise CommandFailure(cmd_args)
+
+        logger.info("%s role setup script(s) completed", framework.name)
+
+def start_test_background(framework, cluster, repo_root):
+    """Start one test on a cluster as a subprocess.
+    The state of these subprocesses lives in the framework objects stored in
+    the fwinfo module
+    """
+    logger.info("Starting cluster configure & test run for %s (will background)",
+                framework.name)
+    _setup_strict(framework, cluster, repo_root)
+
+    logger.info("Launching shakedown for %s", framework.name)
+
+    custom_env = os.environ.copy()
+    custom_env['TEST_GITHUB_LABEL'] = framework.name
+    custom_env['STUB_UNIVERSE_URL'] = framework.stub_universe_url
+    custom_env['CLUSTER_URL'] = cluster.url
+    custom_env['CLUSTER_AUTH_TOKEN'] = cluster.auth_token
+
+    runtests_script = os.path.join(repo_root, 'tools', 'run_tests.py')
```

Comment:
Calling directly into run_tests.py makes a lot of sense.  However, it wants a couple things.

First it needs to be able to run pytest in a detached fashion.  
Next there's a LOT of overlap in concepts at this point between run_tests.py and test.py + venvutil + cli_install 
I'm currently stuck on the idea of a workspace for a given framework test run.  run_tests.py certainly a fairly clean one but insists on wiping it exit, which makes debugging very hard.  Also I'd like the output to be where I can conveniently refer to it from a jenkins run to collect artifacts.  However, developers are usually not going to want to keep a lot of test run temp files accruing.  

These aren't hard problems, I just haven't planned it out.